### PR TITLE
EVM versioning and Paris Updates

### LINKED ISSUE #3882: Low-level calls to invalid addresses revert
### Description

If you call a contract that is not deployed, any call made to it reverts with `INVALID_SOLIDITY_ADDRESS`. 
Low-level calls should not revert when an address does not exist. This could break many contracts ported from Ethereum or other EVM.

For example, following will revert, given `ADDRESS` does not exist:

```solidity
EX1:: ADDRESS.code.length;
EX2:: assembly { extcodesize(ADDRESS) }
EX3:: ADDRESS.call("")
EX4:: ADDRESS.staticcall("")
EX5:: try ISomeInterface(ADDRESS).someFunction() {} catch {}
```

This requires using some middleman contract to check if a contract exists, and call the middleman contract with try/catch or low-level call. It would be an antipattern if any call supposed to not fail had to call a middleman contract first to check if contract exists.

### Steps to reproduce

Deploy and test something like this:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity =0.6.12;

contract ContractSizeChecker {
    function checkCodeSize(address contractAddress) external view returns (uint size) {
        assembly {
            size := extcodesize(contractAddress)
        }
    }
}
```

This call will revert given `contractAddress` does not exist. So this makes using something like [OpenZeppelin's `isContract` function](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/1c3410aa75d58ca3bf4bdc3f3ef8c9d8f8f8a860/contracts/utils/Address.sol#L36-L42) not possible. Instead, one have to deploy `ContractSizeChecker` separately, then make a low-level staticcall to it. That has its other issues as well, so it probably shouldn't be used.

### Additional context

_No response_

### Hedera network

testnet

### Version

n/a

### Operating system

_No response_

### PR DESCRIPTION
**Pull request by shemnon**
Draft impl of EVM versions, including design doc.

Signed-off-by: Danno Ferrin <danno.ferrin@swirldslabs.com>

**Description**:

**Related issue(s)**:

Fixes #3882 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: docs/design/evm-versioning.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+|----------------:|:--------------|:--------------------------------------------------------------------------------------|
+|         `v0.30` | `London`      |                                                                                       |
+|         `v0.31` | `Paris`       | Replaces `DIFFICULTY` with `RANDAO`, removes errors from Invalid Solidity Addresses   |
+|               ? | `Shanghai`    | Expected Q2 2023 or later                                                             |
+
+## Open Questions
+
+The exact timing of versions that correspond to Ethereum Mainnet forks is out of scope of this document
+
+## Acceptance Tests
```
Comment (Nana-EC) [inline docs/design/evm-versioning.md:62]: Should call out a version verification scenarios that should be captured.

**Review by Nana-EC — COMMENTED** [review]
1st pass.
LG so far. Will circle back for 2nd.
Notably design doc should be checked in for 0.30 whiles we hold the code in draft for 0.31

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Nice draft.
I like the direction.
Provided a suggestion on the interfaces and had some clarifying questions.

Please also break this into 2 PR's.
Since most of the comments are code side you can leave the code here and then open a new PR just for the design doc that can go into 0.30.

File: docs/design/evm-versioning.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+|  Hedera Version | Ethereum Fork | Comments                                                                              |
+|----------------:|:--------------|:--------------------------------------------------------------------------------------|
+|         `v0.30` | `London`      |                                                                                       |
+|         `v0.31` | `Paris`       | Replaces `DIFFICULTY` with `RANDAO`, removes errors from Invalid Solidity Addresses   |
+|               ? | `Shanghai`    | Expected Q2 2023 or later                                                             |
+
+## Open Questions
+
+The exact timing of versions that correspond to Ethereum Mainnet forks is out of scope of this document
```
[the remark this answers is not in the document]
Comment (shemnon) [reply docs/design/evm-versioning.md:60]: Generally speaking yes, on mainnet the "transition" is well established a head of time.  Formerly at a block number, now at a timestamp.

File: hedera-node/src/main/java/com/hedera/services/contracts/ContractsModule.java [hunk 3]
```diff
@@ -264,13 +159,27 @@ PrecompiledContract bindExchangeRatePrecompile(
...
+    }
+
+    @Provides
+    @Singleton
+    @IntoMap
+    @StringKey("v0.31")
+    static EVM provideV31EVM(@V_0_31 Set<Operation> hederaOperations, GasCalculator gasCalculator) {
+        var operationRegistry = new OperationRegistry();
+        // ChainID will be overridden
+        registerLondonOperations(operationRegistry, gasCalculator, BigInteger.ZERO);
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/src/main/java/com/hedera/services/contracts/ContractsModule.java]: Moot, as we immediately overwrite it with PRNGSeed.  But I'll update it to be explicit.

File: hedera-node/src/main/java/com/hedera/services/contracts/ContractsV_0_31Operations.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+import java.util.function.BiPredicate;
+import javax.inject.Singleton;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.evm.operation.Operation;
+import org.hyperledger.besu.evm.precompile.PrecompiledContract;
+
+@Module
+public interface ContractsV_0_31Operations {
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/src/main/java/com/hedera/services/contracts/ContractsV_0_31Operations.java:59]: V_0_30 is the earliest version I expect we will need to support with retrospective execution.  Next stop is V22 IIRC.

File: hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java [hunk 5]
```diff
@@ -470,14 +445,9 @@ protected void process(final MessageFrame frame, final OperationTracer operation
...
-                return messageCallProcessor;
-            case CONTRACT_CREATION:
-                return contractCreationProcessor;
-            default:
-                throw new IllegalStateException(
-                        "Request for unsupported message processor type " + type);
-        }
+        return switch (type) {
+            case MESSAGE_CALL -> messageCallProcessor;
+            case CONTRACT_CREATION -> contractCreationProcessor;
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java:435]: new switch expression semantics.  Null throws an exception, hence if all enum cases are taken care of then the default case is dropped as it is dead code.

File: hedera-node/src/main/resources/bootstrap.properties [hunk 6]
```diff
@@ -55,6 +55,7 @@ contracts.allowAutoAssociations=false
 contracts.allowCreate2=true
 contracts.chainId=295
 contracts.defaultLifetime=7890000
+contracts.evm.version=v0.30
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/src/main/resources/bootstrap.properties:58]: Stylistic choice, to make it clear it's a version as opposed to a fork name like "Paris" or "London"

**Review by lukelee-sl — APPROVED** [review]

**Review by Nana-EC — APPROVED** [review]
LG.
Unblocking and provided some improvement suggestions

File: docs/design/evm-versioning.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (shemnon) [reply docs/design/evm-versioning.md:60]: version is a function of block number.  It's just repetitive data that will bloa the record stream if we put it in every transaction record.

File: hedera-node/src/main/java/com/hedera/services/contracts/operation/HederaChainIdOperation.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                new OperationResult(
+                        OptionalLong.of(gasCalculator.getBaseTierGasCost()), Optional.empty());
+        this.oogResponse = new OperationResult(optionalGasCost, Optional.of(INSUFFICIENT_GAS));
+    }
+
+    public Operation.OperationResult execute(final MessageFrame frame, final EVM evm) {
+        if (frame.getRemainingGas() < gasCost) {
+            return oogResponse;
+        }
+        frame.pushStackItem(globalDynamicProperties.chainIdBytes32());
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/contracts/operation/HederaChainIdOperation.java:55]: Much nicer. 😁

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM !

File: docs/design/evm-versioning.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+## Architecture
+
+First, the EVM object will be created by dagger injection. Previously these were constructed in the EVM Processor for
+each HederaTransaction Type, where Dagger provided the operations and gas calculator and the EVM was created in situ.
+
+Second, the dagger injection will be done via a map of providers, keyed by a text string corresponding to the advertised
+version of the EVM. Using a provider will result in just-in-time construction of the object and will not cause multiple
+versions of the EVM to exist in-memory.  (Done via `@IntoMap` bindings for the EVM and injected
+as `Map&lt;String, Provider&lt;EVM>>`).
```
[the remark this answers is not in the document]
Comment (shemnon) [reply docs/design/evm-versioning.md:34]: This is for maximal markdown compatibility.

File: docs/design/evm-versioning.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+Finally, the required differences will be populated into separate submodules, using a version appropriate dagger
+qualifier.
+
+## Non-Functional Requirements
+
+A standard dynamic property will be used to configure the EVM at startup (`contracts.evm.version`) and will be defaulted
+to the last activated version on mainnet in the event the value is not set.
+
+EVM versions will follow the format `v&lt;major>.&lt;minor>`, corresponding to the released version of hedera. for
```
[the remark this answers is not in the document]
Comment (shemnon) [reply docs/design/evm-versioning.md:44]: won't fix, this is for maximal markdown compatibility.

File: docs/design/evm-versioning.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+example, the earliest version supported by this regime is `v0.30` with a planned update for `v0.31`. Not every hedera
+version will have a new EVM version. Only when EVM compatability is impacted for object replay will a new version be
+set.
+
+It is expected that there will be a new version for each major Ethereum Mainnet hard fork. A table will be kept here to
+document which major hardfork corresponds to each internal version.
+
+|  Hedera Version | Ethereum Fork | Comments                                                                              |
```
[the remark this answers is not in the document]
Comment (shemnon) [reply docs/design/evm-versioning.md:52]: Linked fork name to Eth specs.

File: hedera-node/src/test/java/com/hedera/services/contracts/execution/CallLocalEvmTxProcessorTest.java [hunk 11]
```diff
@@ -81,21 +86,29 @@ class CallLocalEvmTxProcessorTest {
...
 
     @BeforeEach
-    private void setup() {
+    public void setup() {
         CommonProcessorSetup.setup(gasCalculator);
 
+        var operationRegistry = new OperationRegistry();
+        MainnetEVMs.registerLondonOperations(operationRegistry, gasCalculator, BigInteger.ZERO);
+        operations.forEach(operationRegistry::put);
+        when(globalDynamicProperties.evmVersion()).thenReturn("v0.30");
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/src/test/java/com/hedera/services/contracts/execution/CallLocalEvmTxProcessorTest.java:103]: Added a field to the dagger modules.

File: hedera-node/src/test/java/com/hedera/services/contracts/ContractsModuleTest.java [hunk 12]
```diff
@@ -29,11 +29,11 @@ class ContractsModuleTest {
 
     @Test
     void logOperationsAreProvided() {
-        final var log0 = ContractsModule.provideLog0Operation(gasCalculator);
-        final var log1 = ContractsModule.provideLog1Operation(gasCalculator);
-        final var log2 = ContractsModule.provideLog2Operation(gasCalculator);
-        final var log3 = ContractsModule.provideLog3Operation(gasCalculator);
-        final var log4 = ContractsModule.provideLog4Operation(gasCalculator);
+        final var log0 = ContractsV_0_30Operations.provideLog0Operation(gasCalculator);
```
[the remark this answers is not in the document]
Comment (shemnon) [reply hedera-node/src/test/java/com/hedera/services/contracts/ContractsModuleTest.java]: Rolled into old test cases.  For tests across all versions i iterate over the injection map.

File: docs/design/evm-versioning.md (hunk 1 above)
Comment (shemnon) [reply docs/design/evm-versioning.md:62]: updated.

**Review by Nana-EC — DISMISSED** [review]
LGTM

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM !

**Comment by Shungy** [comment]
Any estimates when this would go live for testnet?

**Comment by bmino** [comment]
also looking for this go-live :)

**Pull request merged**